packages/jupyter-ai/src/components/chat-settings.tsx

```diff
@@ -327,6 +363,87 @@ export function ChatSettings(props: ChatSettingsProps): JSX.Element {
         )}
       </Select>
 
+      {/* Completer language model section */}
+      <h2 className="jp-ai-ChatSettings-header">
+        Completer model
+        {props.completionProvider ? (
+          <UseSignal signal={props.completionProvider.settingsChanged}>
+            {(): JSX.Element => (
+              <CompleterSettingsButton
+                provider={props.completionProvider}
+                openSettings={props.openInlineCompleterSettings}
+                selection={clmProvider}
+              />
+            )}
+          </UseSignal>
+        ) : (
+          <CompleterSettingsButton
+            provider={null}
+            openSettings={props.openInlineCompleterSettings}
+            selection={clmProvider}
+          />
+        )}
```

Comment:
1. The reason the ternary expression is used is because you cannot access `props.completionProvider.settingsChanged` signal if `props.completionProvider` is `null`. Because `completionProvider` comes from a separate plugin, which can be disabled, it can be null. If there is no signal, then there is no reason and no way to wrap it in `UseSignal` component.
2. Yes, we can move the `UseSignal` down to `CompleterSettingsButton`.

What is your concern here? Assuming that you just want to push the logic out of the `ChatSettings` components which is getting big, I went ahead and moved the use signal down to `CompleterSettingsButton` implementation in 8d727d38407d59e73db78a1482992d65e18760e1 and then made use of caching the enabled state (from 13d0ecf13417c07139f8031e18dca080e2ca0f94) in 92fa8c6d9d6ffdbc050bcb78a508ce757a4731ed which removed the `UseSignal`.

Reply:
There are two main reasons I prefer avoiding HOCs:

1. HOCs are some significant performance penalties in some scenarios. In the original code block I highlighted here, the child component of this HOC is re-mounted on every update rather than re-rendered, since a new anonymous component `() => ( ... )` is being defined every time `props.completionProvider` is `true` when the expression is re-evaluated on re-render. 
2. React Hooks were intended to replace HOCs, and as a result, newer contributors may be less familiar with this pattern. To be honest, I haven't encountered an HOC in quite some time, so it took me a while to understand what was happening in the original code block I highlighted here.